hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/RatisUnderReplicationHandler.java

```diff
@@ -137,6 +150,73 @@ public int processAndSendCommands(
     return commandsSent;
   }
 
+  private void removeUnhealthyReplicaIfPossible(ContainerInfo containerInfo,
+      Set<ContainerReplica> replicas, List<ContainerReplicaOp> pendingOps)
+      throws NotLeaderException {
+    for (ContainerReplicaOp op : pendingOps) {
+      if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+        // There is at least one pending delete which will free up a node.
+        // Therefore we do nothing until that delete completes or times out.
+        LOG.debug("Container {} has pending deletes which will free nodes",
+            pendingOps);
+        return;
+      }
+    }
+
+    if (replicas.size() <= 2) {
+      // We never remove replicas if it will leave us with less than 2 replicas
+      LOG.debug("There are only {} replicas for container {} so no more will " +
+          "be deleted", replicas.size(), containerInfo);
+      return;
+    }
+
+    // If there are no healthy containers, we do nothing as this should be
+    // handled elsewhere.
+    boolean hasHealthy = false;
+    for (ContainerReplica r : replicas) {
+      if (r.getState() == State.CLOSED) {
```

Comment:
I don't think I understand what you are suggesting. The code only gets here if we cannot find a spare node, so that cover the "more space needed" check.

Then we prefer deleting any quasi-closed with a lagging sequence if any exists.

Only then do we delete a single unhealthy if there are no quasi-closed to remove.

Are you saying that we should check for the CONTAINER being quasi-closed, and if the container is quasi-closed, we should delete unhealthy over quasi-closed?

Reply:
> Are you saying that we should check for the CONTAINER being quasi-closed, and if the container is quasi-closed, we should delete unhealthy over quasi-closed?

Yes. We can be in the following state in a 4 node cluster:
SCM: Quasi-closed
DNs: quasi (origin 1), unhealthy (origin 2), unhealthy (origin 3), quasi (origin 1)
Ideally we would keep the unhealthy containers because they have unique origin nodes. This is what the new and old RMs currently do. However in this case we need to remove one of the unhealthy replicas to restore the quasi-closed container to full replication.